src/nccl_ofi_rdma.c

```diff
@@ -6842,13 +6870,20 @@ static int init_rail_ofi_resources(nccl_net_ofi_rdma_device_t *device,
 	nccl_net_ofi_rdma_device_rail_t *rail_dev;
 	nccl_net_ofi_ep_rail_t *rail;
 	nccl_net_ofi_ep_rail_t *control_rail;
+	struct fid_domain *domain;
+	struct fid_cq *cq;
 
 	/* Initialize libfabric resources of endpoint rails */
 	for (int rail_id = 0; rail_id != ep->num_rails; ++rail_id) {
 		rail_dev = rdma_device_get_rail(device, rail_id);
 		rail = rdma_endpoint_get_rail(ep, rail_id);
 
-		ret = ep_rail_init(ep, dev_id, rail_id, rail_dev, rail);
+		ret = get_domain_and_cq_for_endpoint_rail(ep, rail_dev, rail, &domain, &cq);
```

Comment:
I think this is confusing -- you're calling this function with `rail`, which is uninitialized at this point (well, it will be all zeros only because we calloc the array..)

I would prefer a solution where we don't pass an uninitialized rail here. Perhaps splitting `get_domain_and_cq_for_endpoint_rail` into two functions, 1) the rail has not been initialized yet, in which case we don't need to pass rail at all, and 2) rail has been initialized, which would really be just the body of the first `if` statement of `get_domain_and_cq_for_endpoint_rail`.